enterprise/coderd/schedule/template.go

```diff
@@ -304,6 +304,23 @@ func (s *EnterpriseTemplateScheduleStore) updateWorkspaceBuild(ctx context.Conte
 		return xerrors.Errorf("calculate new autostop for workspace %q: %w", workspace.ID, err)
 	}
 
+	if workspace.AutostartSchedule.Valid {
+		templateScheduleOptions, err := s.Get(ctx, db, workspace.TemplateID)
+		if err != nil {
+			return xerrors.Errorf("get template schedule options: %w", err)
+		}
+
+		nextStartAt, _ := agpl.NextAutostart(dbtime.Now(), workspace.AutostartSchedule.String, templateScheduleOptions)
+
+		err = db.UpdateWorkspaceNextStartAt(ctx, database.UpdateWorkspaceNextStartAtParams{
+			ID:          workspace.ID,
+			NextStartAt: sql.NullTime{Valid: true, Time: nextStartAt},
+		})
```

Comment:
Just to check I understand. You're suggesting we have a `TRIGGER` on certain conditions (those that invalidate our `next_start_at`) so we can keep `next_start_at` in sync? 

As the logic for calculating the `next_start_at` is easier to run on `coderd` - due to a lack of builtin support for cron expressions in Postgres - I assume this `TRIGGER` would `NOTIFY` our `coderd` server to perform this update?

My understanding (and please correct me if I'm wrong) is that `NOTIFY` doesn't _guarantee_ delivery, so could it be possible that we fail to update a `next_start_at` with this approach? If so, is the likelihood of this small enough to not be a concern?

Reply:
(Side note: One risk in the trigger approach is that unless our relationships between tables is super clear (i.e. require one vs multiple joins), it could add a lot of overhead to updating the tables that have triggers.)

Reply:
> This would force the Go code to re-fetch the row on next iteration and recompute the new start?

Yeah this is definitely an option we could look at doing

Reply:
I forgot to respond yesterday :facepalm: 

@mafredri yes this is correct. Just using `TRIGGER` to invalidate the column (set to null).

At the moment we add the invalidation to the http routes that affect it. That just feels a bit leaky and might be a bug down the road when we forget to invalidate it.

Reply:
I've got this working!

When `autostart_schedule` is updated but `next_start_at` _isn't_ updated, the database sets `next_start_at` to `NULL`. On the next run of the lifecycle executor, it sees that the column is `NULL`, so it computes it and updates it in the database.

```
coder=# select name, autostart_schedule, next_start_at from workspaces;
     name     |    autostart_schedule    |     next_start_at
--------------+--------------------------+------------------------
 teal-bass-93 | CRON_TZ=UTC 0 13 * * 1-5 | 2024-11-28 13:00:00+00
(1 row)

coder=# update workspaces set autostart_schedule='CRON_TZ=UTC 30 14 * * 1-5' where name='teal-bass-93';
UPDATE 1

coder=# select name, autostart_schedule, next_start_at from workspaces;
     name     |    autostart_schedule     | next_start_at
--------------+---------------------------+---------------
 teal-bass-93 | CRON_TZ=UTC 30 14 * * 1-5 |
(1 row)

... a minute later ...

coder=# select name, autostart_schedule, next_start_at from workspaces;
     name     |    autostart_schedule     |     next_start_at
--------------+---------------------------+------------------------
 teal-bass-93 | CRON_TZ=UTC 30 14 * * 1-5 | 2024-11-28 14:30:00+00
(1 row)
```

Reply:
I think this will work nicely!